### PR TITLE
[Enhancement] Make local partition topn streaming

### PR DESCRIPTION
**Pull request by ZiheLiu**
Signed-off-by: ZiheLiu <ziheliu1024@gmail.com>

## What type of PR is this：
- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Which issues of this PR fixes ：

Relates to #17553.

## Problem Summary(Required) ：

Before this PR, the logic of local partition top n is as follows:
1. Add all the chunks to `PartitionHashMap`, where key is partition columns and value is chunks in each partition.
2. Sort each partition chunks.

After this PR, we sort partition chunks when the partition has received enough num rows (4096).

## Modification
Add two arguments to  `PartitionHashMap::append_chunk`:
- `new_partition_cb`: `void(size_t partition_idx)`
    called when coming a new key not in the hash map.
- `partition_chunk_consumer`: `void(size_t partition_idx, const ChunkPtr& chunk)`
    called for each partition with enough num rows (4096) after adding chunk to the hash map.

### Environment
3 BE (16vCPUs, 64GB RAM)
ssb 100g

### Queries
```sql
-- Q1 low cardinal (63), HashMapWithSerializedKeyFixedSize
with w1 as (select lo_tax, lo_linenumber, row_number() over(partition by lo_tax, lo_linenumber order by lo_orderkey) as rn from lineorder_flat) 
select 
    ifnull(sum(murmur_hash3_32(lo_tax)), 0)
    + ifnull(sum(murmur_hash3_32(lo_linenumber)), 0)
    + ifnull(sum(murmur_hash3_32(rn)), 0) 
from w1 where rn = 1


-- Q2 low cardinal (225), HashMapWithSerializedKey
with w1 as (select lo_tax, s_nation, row_number() over(partition by lo_tax, s_nation order by lo_orderkey) as rn from lineorder_flat) 
select /*+ SET_VAR(cbo_enable_low_cardinality_optimize=false) */ 
    ifnull(sum(murmur_hash3_32(lo_tax)), 0)
    + ifnull(sum(murmur_hash3_32(s_nation)), 0)
    + ifnull(sum(murmur_hash3_32(rn)), 0)
from w1 where rn = 1

-- Q3 high cardinal (13998451), HashMapWithSerializedKeyFixedSize
with w1 as (select lo_custkey, lo_linenumber, row_number() over(partition by lo_custkey, lo_linenumber order by lo_orderkey) as rn from lineorder_flat) 
select 
    ifnull(sum(murmur_hash3_32(lo_custkey)), 0)
    + ifnull(sum(murmur_hash3_32(lo_linenumber)), 0)
    + ifnull(sum(murmur_hash3_32(rn)), 0) 
from w1 where rn = 1

-- Q4 high cardinal (49984917), HashMapWithSerializedKey
with w1 as (select c_name, s_nation, row_number() over(partition by c_name, s_nation order by lo_orderkey) as rn from lineorder_flat) 
select /*+ SET_VAR(cbo_enable_low_cardinality_optimize=false) */ 
    ifnull(sum(murmur_hash3_32(c_name)), 0)
    + ifnull(sum(murmur_hash3_32(s_nation)), 0)
    + ifnull(sum(murmur_hash3_32(rn)), 0) 
from w1 where rn = 1


-- Q5 very high cardinal (600037902), HashMapWithSerializedKeyFixedSize
with w1 as (select lo_orderkey, lo_linenumber, row_number() over(partition by lo_orderkey, lo_linenumber order by lo_orderkey) as rn from lineorder_flat) 
select
    ifnull(sum(murmur_hash3_32(lo_orderkey)), 0)
    + ifnull(sum(murmur_hash3_32(lo_linenumber)), 0)
    + ifnull(sum(murmur_hash3_32(rn)), 0) 
from w1 where rn = 1


-- Q6 very high  cardinal (150000000), HashMapWithSerializedKey
with w1 as (select lo_orderkey, c_name,  row_number() over(partition by lo_orderkey, c_name order by lo_orderkey) as rn from lineorder_flat) 
select /*+ SET_VAR(cbo_enable_low_cardinality_optimize=false) */ 
    ifnull(sum(murmur_hash3_32(lo_orderkey)), 0)
    + ifnull(sum(murmur_hash3_32(c_name)), 0)
    + ifnull(sum(murmur_hash3_32(rn)), 0) 
from w1 where rn = 1

```

### Result
| Name | Time (s)       | Peak Memory (GB) | Time (s)                 | Peak Memory (GB) |
| ---- | -------------- | ---------------- | ------------------------ | ---------------- |
|      | MultiPartition |                  | MultiPartition-Streaming |                  |
| Q1   | 1.24           | 1.84             | 1.07                     | **0.03**             |
| Q2   | 4.49           | 7.65             | 3.81                     | **0.20**             |
| Q3   | 11.11          | 9.13             | 11.50                    | 9.43             |
| Q4   | OOM            | -                | OOM                      | -                |
| Q5   | 7.96           | 5.09             | 8.04                     | 4.99             |
| Q6   |12.39          | 23.78            | 12.09                    | 23.85            |

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto backported to target branch
  - [x] 2.5
  - [ ] 2.4
  - [ ] 2.3
  - [ ] 2.2

### CODE REVIEW

**Comment by ZiheLiu** [comment]
run starrocks_admit_test

**Comment by ZiheLiu** [comment]
run starrocks_admit_test

**Review by satanson — APPROVED** [review]

File: be/src/exec/partition/chunks_partitioner.h [hunk 1]
```diff
@@ -182,7 +215,7 @@ class ChunksPartitioner {
         while (chunk_it != chunk_end) {
             // Because we first fetch chunks from hash_map, so the _partition_idx here
             // is already set to hash_map.size()
-            if (!consumer(_partition_idx, *chunk_it++)) {
```
Comment (liuyehcf) [inline be/src/exec/partition/chunks_partitioner.h:185]: before: _partition_idx = sizeof(keys)
after: hash_map_with_key.kNullKeyPartitionIdx = 0
is that what you expected?

---

I see it

```cpp
            // partition_idx=0 is reserved by null key.
            auto next_partition_idx = hash_map.size() + 1;
```

File: be/src/exec/partition/partition_hash_map.h [hunk 2]
```diff
@@ -114,32 +144,48 @@ struct PartitionHashMapBase {
...
+                return ctor(key_allocator(key), part_chunks);
             });
             if (is_new_partition) {
                 check_downgrade(hash_map);
+                new_partition_cb(next_partition_idx++);
             }
+
             auto& value = *(iter->second);
             if (value.chunks.empty() || value.remain_size <= 0) {
-                if (!value.chunks.empty()) {
```
Comment (liuyehcf) [inline be/src/exec/partition/partition_hash_map.h:142]: what's the new condition means?

**Review by kangkaisen — APPROVED** [review]

**Review by liuyehcf — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

File: be/src/exec/partition/partition_hash_map.h (hunk 2 above)
Comment (ZiheLiu) [reply be/src/exec/partition/partition_hash_map.h:142]: If `value.select_indexes` is empty, then we also needn't call `append_selective`.

**Comment by ZiheLiu** [comment]
run starrocks_be_unittest

**Review by Seaven — APPROVED** [review]

**Pull request merged**